hack/generator/pkg/functions/property_assignment_function.go

```diff
@@ -198,38 +199,97 @@ func (fn *PropertyAssignmentFunction) generateBody(
-	}
+	// source is the identifier from which we are reading values
+	source := fn.direction.SelectString(parameter, receiver)
+
+	// destination is the identifier onto which we write values
+	destination := fn.direction.SelectString(receiver, parameter)
+
+	bagPrologue := fn.createPropertyBagPrologue(source, generationContext)
+	assignments := fn.generateAssignments(dst.NewIdent(source), dst.NewIdent(destination), generationContext)
+	bagEpilogue := fn.propertyBagEpilogue(destination)
+
+	return astbuilder.Statements(
+		bagPrologue,
+		assignments,
+		bagEpilogue,
+		astbuilder.ReturnNoError())
 }
 
-// generateDirectConversionFrom returns the method body required to directly copy information from
-// the parameter instance onto our receiver
-func (fn *PropertyAssignmentFunction) generateDirectConversionFrom(
-	receiver string,
-	parameter string,
-	generationContext *astmodel.CodeGenerationContext,
-) []dst.Stmt {
-	result := fn.generateAssignments(dst.NewIdent(parameter), dst.NewIdent(receiver), generationContext)
-	result = append(result, astbuilder.ReturnNoError())
-	return result
+// createPropertyBagPrologue creates any introductory statements needed to set up our property bag before we start doing
+// assignments. We need to handle three cases:
+//   o If our source has a property bag, we clone it.
+//   o If our destination has a property bag (and our source does not), we create a new one.
+//   o If neither source nor destination has a property bag, we don't need to do anything.
+// source is the name of the source to read the property bag from
+func (fn *PropertyAssignmentFunction) createPropertyBagPrologue(
+	source string,
+	generationContext *astmodel.CodeGenerationContext) []dst.Stmt {
+
+	sourcePropertyBag, sourcePropertyBagFound := fn.findPropertyBagProperty(fn.sourceType())
+	_, destinationPropertyBagFound := fn.findPropertyBagProperty(fn.destinationType())
+
+	// If we're not using the property bag, don't declare one
+	// We're using it if we are
+	// (a) reading from it; or
+	// (b) writing to it; or
+	// (c) we need one to store in the final object
+	//
+	if !fn.readsFromPropertyBag && !fn.writesToPropertyBag && !destinationPropertyBagFound {
+		return nil
+	}
+
+	// Don't refactor the local genruntimePkg out to this scope - calling MustGetImportedPackageName() flags the
+	// package as referenced, so we must only call that if we are actually going to reference the genruntime package
+
+	if sourcePropertyBagFound {
+		// Found a property bag on our source type, need to clone it to allow removal of values
+		genruntimePkg := generationContext.MustGetImportedPackageName(astmodel.GenRuntimeReference)
+		cloneBag := astbuilder.ShortDeclaration(
+			fn.conversionContext.PropertyBagName(),
+			astbuilder.CallQualifiedFunc(
+				genruntimePkg,
+				"NewPropertyBag",
+				astbuilder.Selector(dst.NewIdent(source), string(sourcePropertyBag.PropertyName()))))
+		cloneBag.Decs.Before = dst.NewLine
```

Comment:
Just confirming - getting a value from a property bag removes it from the bag doesn't it? I haven't seen anything that does that explicitly in the golden files, so it would need to happen on the pull.

Reply:
Yes, `PropertyBag.Pull()` is destructive. It's documented as 
```
// Pull removes a value from the bag, using it to populate the destination
```